### ISSUE TITLE
error running "node ./install-eaf-win32.js"

### ISSUE THREAD

**Issue by mikereape** [manateelazycat/emacs-application-framework#555]
I am trying to install eaf on a 64-bit Windows 10 machine. The Emacs I use is downloaded from the usual place. I got an .msi installer so it behaves much better than the old way with no installer. The version is "GNU Emacs 27.1 (build 1, x86_64-w64-mingw32)" of 2020-08-21. I installed nodejs and npm from scratch. It downloaded some Visual Studio libraries to compile some C code. That all went flawlessly.

The problem is clear to see in the attached log. The error message is identical if I execute "install-eaf-win32.js" in Powershell after cd'ing to `~/.emacs.d/site-lisp/emacs-application-framework/.` The log attached is the result of doing `M-x eaf-install-dependencies` in Emacs. The command to install the dependencies from within Emacs is not `install-eaf-dependencies`. It is `eaf-install-dependencies` There was a bunch of problems with elpa not finding packages in particular getting ctables and epc installed was quite difficult. I ended up using a combination of el-get and package-install and package-install-file. To get the log shown by doing `M-x eaf-install-dependencies` I  I had to copy `install-eaf-win32.js` into my `~/.emacs.d/elpa/eaf-0.5/app/` because I was getting error messages telling me it couldn't find the .js file in that directory.

[log4.log](https://github.com/manateelazycat/emacs-application-framework/files/5901284/log4.log)

I got the exact same error and stack trace repeatedly in Powershell and within Emacs and even in Git Bash for a while before it mysteriously stopped seeing nodejs.

The expected behaviour is that the dependencies should just get installed without the NOENT error.

I used the command you listed in step 1 to download eaf:

`git clone --depth=1 -b master https://github.com/manateelazycat/emacs-application-framework.git ~/.emacs.d/site-lisp/emacs-application-framework/`

so the version is whatever it was just before 1200AM GMT on 2021-01-31.

That about covers it I think. I very much look forward to your reply. I think this is awesome what you've put together. I can't wait to get it working and start using it. I have Win 10 and Ubuntu Server machines. I need eaf on all of them. I would love to run it on my Mac Mini as well. Is your Docker support an option for that? Just run Emacs in Docker in macOS?

Thanks very much in advance
Best, Mike

**Comment by MatthewZMD** [comment]
Unfortunately I don't use Windows and never seen this issue on Linux before, we need more Windows users to assist you with this problem.

**Comment by vconcat** [comment]
From the log, I think the problem is `Error: spawnSync C:\WINDOWS\system32\cmd.exe ENOENT`.
Search with keywords `spawnSync cmd.exe ENOENT`, and there are many results about it.

For example, https://github.com/AtomLinter/linter-tidy/issues/68

**Comment by mikereape** [comment]
Hi okay so I updated everything as requested. I followed the instructions here: `https://nodejs.dev/learn/update-all-the-nodejs-dependencies-to-their-latest-version/`. I have attached another log `log5.` It's the same as the previous one once you get past the parts that are timing-dependent. However if you look at the file `JScript error.log` you can see that the Windows Script Host is complaining about the first byte in the file. No idea.

[JScript error.log](https://github.com/manateelazycat/emacs-application-framework/files/5908961/JScript.error.log)
[log5.log](https://github.com/manateelazycat/emacs-application-framework/files/5908963/log5.log)

I tried a couple things I found online like running it as `node install-eaf-win32.js` and not just bare `install-eaf-win32.js` in Powershell and got the same problem as previous in Powershell. And also got the same problem as last time as reflected in the logs when I re-tried `eaf-install-dependencies` in Emacs.

The method that I used to install as I said last time was to download a `.msi` installer off the Node.js site here: `https://nodejs.org/en/download/`. I just grabbed the 64-bit installer.

Again, just to make sure I hadn't messed up I re-started from scratch again. The installer uses chocolatey to install a bunch of stuff including Visual Studio 2017 libraries to compile C code. chocolatey is great when it works. When it doesn't you are very _very_ lucky if you can clean up after it. It took a long time and a lot of effort. Ironically, to install node it installs Python! I'm not gonna do it again.

I can _read_ javascript but that's about it. And since there is no js standard or reference implementation who knows what the Windows Script Host expects.

I would suggest that for development if you don't have access to a Win 10 box that you grab a copy of Win 10 off the MS download site and run it under virtualbox. There are lots of blog posts about how to do this including at the How-To Geek site. It'll run even without a license key they say. (I haven't tried it myself.) I would start with this article: `https://www.howtogeek.com/244678/you-dont-need-a-product-key-to-install-and-use-windows-10/`

I really think this is a great project and I have a use for it _right now_ but I am not the right person to be sorting out arcane javascript problems. (It makes me wonder why you didn't use Python to install the deps.) It does seem to me (no retaliatory flames please I don't deserve them) that to say it runs on Windows is currently a bit of stretch if the core developers haven't actually tested it on Windows. Anyone out there reading this issue thread that has it running on Win 10 (Pro)?

Anxiously anticipating
Mike

**Comment by MatthewZMD** [comment]
> However if you look at the file JScript error.log you can see that the Windows Script Host is complaining about the first byte in the file. No idea.

Why is Windows running JScript instead of JavaScript... How did you trigger this error? Did you do something different than last time when you initially made this post?

> Windows is currently a bit of stretch if the core developers haven't actually tested it on Windows.

This is a community-driven project. The Windows port of this project is implemented by active members of the community through PRs and tested by people in the community, and they said it works for them. There are occasional issues with some people and resulted in issues like this one here, and sometimes is really needing the community's help in diagnosing the problem.  Aside from that, what you experienced here is not unable to run on Windows, but unable to install some npm dependencies, it's more of an npm issue than an issue about EAF itself.

**Comment by MatthewZMD** [comment]
Could you `git pull` and run the script again, this is a workaround that shouldn't stop the script from running if there is an error, the Python dependencies should be installed normally.

**Comment by mikereape** [comment]
Hi I did what you asked. Here's what happened.

[log8.log](https://github.com/manateelazycat/emacs-application-framework/files/5913368/log8.log)

Once you get to the line that says `found 0 vulnerabilities` it's identical from there to the bottom:

FWIW, I also got this when I did the pull:

[git pull 2021-02-02.log](https://github.com/manateelazycat/emacs-application-framework/files/5913390/git.pull.2021-02-02.log)

I also ran it in Powershell after that just to be sure and got this:

[powershell log 2021-02-02.log](https://github.com/manateelazycat/emacs-application-framework/files/5913465/powershell.log.2021-02-02.log)

Once again, once you get to the line that says `found 0 vulnerabilities` it's identical from there to the bottom. JScript is just MS's name for Javascript. The Windows Script Host takes care of executing scripts that are just called by name. It doesn't work like in Linux. JScript is _not_ identical to Ecmascript. Even if you just start with the Wikipedia article on JScript you'll fairly quickly get to these two urls:

`https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2010/894hfyb4(v=vs.100)?redirectedfrom=MSDN`

`https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2010/x85xxsf4(v=vs.100)`

They are both dated 10/04/2012. The Wikipedia article says that the last stable release was version 9 in March, 2011. You'll remember that the error I get when I do it in Powershell and don't put `node` before the js installer file is "Microsoft JScript Compilation Error". That is one of the differences between JScript and ECMAScript. Not only can the Windows Script Host compile JScript files but it can do so conditionally. It's part of the semantics of JScript. And there are other differences. But like I said previously, I can just about read modern (plain vanilla) Javascript and understand it and that's about it.

I really sincerely and humbly but seriously suggest you spin up a Win 10 Pro virtual machine any way you like and try this. I am quite happy to admit that this is at the limits of what I can do. For all I know I'm blissfully unaware that I'm completely missing the complelely obvious. The last time I got paid for writing Javascript was over 20 years ago and as fate would have it it was MS's version and _that_ was _very different_ from the ECMAScript standard of the day. We are talking of the MS of days of old. And 2011 is still old-school MS.

I want to thank you all for all your efforts so far and for sticking with me up to now and for your patience.

Again I anxiously anticipate. Mike

**Comment by mikereape** [comment]
FWIW, except for a couple of years when I was contractually prevented by an NDA, all the code I've ever written was freely available to whoever wanted it.

**Comment by MatthewZMD** [comment]
Seems like your

> I also ran it in Powershell after that just to be sure and got this:
> 
> [powershell log 2021-02-02.log](https://github.com/manateelazycat/emacs-application-framework/files/5913465/powershell.log.2021-02-02.log)

**Comment by mikereape** [comment]
Hi I think I know what the issue is with the Python dependencies. It has to do with my local environment and the fact that the Node .msi installer insisted on installing Python (which I didn't want it to do). It's too long and complicated to explain. I will try pip install or equivalent as you suggest but it may be a day or so before I get around to it.

By all means, do try a VM as you say. The links I supplied previously should get you going in particular the How-To Geek ones. There's lots of stuff "out there" about how you can do this. Feel free to ping me for more pointers to blog posts, articles, etc. including directly by email if you want to keep this issue clean of the details and mechanics of that. If I have time (unlikely I have to admit) I might try it myself on a local Linux box.

I _just can't wait_ for this all to be working. Like a wee kiddie _I'm very excited with anticipation_.

Thanks very much for everything so far and in advance
Mike

**Comment by hjtoh** [comment]
I got the same problem today and I don't think it's a python problem because I installed eaf one week ago without any problem. I think maybe some changes the author made recently caused the problem.

**Comment by wsw0108** [comment]
`mermaid` removed, directory `app/mermaid` no longer exists.

@mikereape @hjtoh

**Comment by MatthewZMD** [comment]
Could you try with the latest patch

**Comment by mikereape** [comment]
Hi

Okay so I tried again. To make sure I did it right, I just did the usual `git clone --depth=1 -b master https://github.com/manateelazycat/emacs-application-framework.git ~/.emacs.d/site-lisp/emacs-application-framework/` and then `eaf-install-dependencies` from within Emacs and got the log attached.

[log9.log](https://github.com/manateelazycat/emacs-application-framework/files/5938974/log9.log)

If I'm reading this right it looks to me like `eaf-install-dependencies` is trying to run `cmd.exe` and is also trying to run `npm install.` As I explained at length, `node.js` and `npm` are already both installed using the recommended installer for Windows 64 bit.

`cmd.exe` isn't going to see it though. `node.js` is either in my system path or my user path and `npm` is in the other. But `cmd.exe` will never see it. It never sees anything. You always have to give it a path to run most executables.

I've looked into the whole issue of the proper way to get Javascript going in _lots and lots and lots_ of detail since last update and have looked at loads of other solutions like Indium, the spacemacs JS layer, Elnode, skeeto's skewer-mode and many, many others.

I've become convinced that the only way to ever get this to work robustly is to install `emacs-lsp` for Typescript. This will run MS `tsserver.` This has _lots_ of advantages. The relevant package is "Tide". The url on GitHub is `https://github.com/ananthakumaran/tide.`

This is MS-sanctioned and circumvents all the JScript/Windows Script host stuff. It has lots of other advantges. I don't have time to pursue this right now but there looks to me like a very simple "trick" to make it easy to do.

You install the relevant plugin into Visual Studio Code and get it to get `tsserver` going ("typescript server" which Tide uses). Then you can connect to it from anything else (like Emacs) at the cost of just starting VS Code first. Since it's VS Code, it's _guaranteed_ to work.

I've become convinced that anything else is, ultimately, a waste of time. Something else might work but this is the way MS intends you to do these things. Since Typescript is a proper superset of Javascript (so MS has always claimed) you can just throw straight Javascript at it and it should fly.

Apologies for the very opinionated post. I do very much like the look of Indium and the spacemas JS layer but I'm a little skeptical about what they use for the backend. I've put a _lot_ of work into all this when I should have been doing lots else. I intend to do this eventually when I can get to it. Unless you fine gentlement sort out the details first :). I wouldn't mind that at all.

Still anxiously awaiting. Oh my!

**Comment by MatthewZMD** [comment]
> I've looked into the whole issue of the proper way to get Javascript going in _lots and lots and lots_ of detail since last update and have looked at loads of other solutions like Indium, the spacemacs JS layer, Elnode, skeeto's skewer-mode and many, many others.
> 
> I've become convinced that the only way to ever get this to work robustly is to install `emacs-lsp` for Typescript. This will run MS `tsserver.` This has _lots_ of advantages. The relevant package is "Tide". The url on GitHub is `https://github.com/ananthakumaran/tide.`
> 
> This is MS-sanctioned and circumvents all the JScript/Windows Script host stuff. It has lots of other advantges. I don't have time to pursue this right now but there looks to me like a very simple "trick" to make it easy to do.
> 
> You install the relevant plugin into Visual Studio Code and get it to get `tsserver` going ("typescript server" which Tide uses). Then you can connect to it from anything else (like Emacs) at the cost of just starting VS Code first. Since it's VS Code, it's _guaranteed_ to work.
> 
> I've become convinced that anything else is, ultimately, a waste of time. Something else might work but this is the way MS intends you to do these things. Since Typescript is a proper superset of Javascript (so MS has always claimed) you can just throw straight Javascript at it and it should fly.
> 
> Apologies for the very opinionated post. I do very much like the look of Indium and the spacemas JS layer but I'm a little skeptical about what they use for the backend. I've put a _lot_ of work into all this when I should have been doing lots else. I intend to do this eventually when I can get to it. Unless you fine gentlement sort out the details first :). I wouldn't mind that at all.
> 
> Still anxiously awaiting. Oh my!

Wait! you don't need any of these, you don't need the language server at all, it's for code completion, never intended for npm dependencies. Oh no.. you got yourself convinced incorrectly.

You're in the wrong direction, you don't need to look at the spacemacs js layer, you don't need to look at how emacs handles javascript, this is completely irrelevent. Your error has nothing to do with emacs.


What you should do right now, is open cmd at the EAF's path and simply run `npm install`, don't waste your time researching anything else, just come back and tell us results :)

**Comment by mikereape** [comment]
I've already explained to you. It's already installed. node.js and npm both.

**Comment by manateelazycat** [comment]
Sorry, we can't help you depend on your explain.

Welcome to leave message if some window user can help.

**Comment by mikereape** [comment]
Three worked. It looks like the terminal one didn't(?):

[npm install x1 2021-02-07.log](https://github.com/manateelazycat/emacs-application-framework/files/5939895/npm.install.x1.2021-02-07.log)
[2021-02-07T20_19_11_003Z-debug.log](https://github.com/manateelazycat/emacs-application-framework/files/5939896/2021-02-07T20_19_11_003Z-debug.log)

Mike

**Comment by MatthewZMD** [comment]
The `npm install` error should hopefully be fixed with the latest commit. Could you please do a `git pull` and try both the `npm install` and `node ./install-eaf-win32.js` to see if both of these commands work for you?

**Comment by mikereape** [comment]
Hi I did do what you asked me to several hours ago but then had to attend to other things. I'm in GMT time but I will endeavour to to forward you the results before I try to get some sleep. It's a bit complicated and I don't want to get it wrong. I thank you in advance for your continue patience.

Mike

**Comment by mikereape** [comment]
Hi I'm back

So these are my three logs from this morning:

This is what I got when I tried to do it in Powershell:
[powershell 2021-02-08.log](https://github.com/manateelazycat/emacs-application-framework/files/5947461/powershell.2021-02-08.log)

which references:

[2021-02-08T06_51_46_829Z-debug.log](https://github.com/manateelazycat/emacs-application-framework/files/5947476/2021-02-08T06_51_46_829Z-debug.log)

This is what I got when I tried to do it via `eaf-install-dependencies` in Emacs:
[eaf-install-dependencies 2021-02-08.log](https://github.com/manateelazycat/emacs-application-framework/files/5947471/eaf-install-dependencies.2021-02-08.log)

However, and I could be very wrong, it looks to me like from a bunch of other GitHub repos that on Windows you have to install build dependencies, etc. The culprit from the logs (looks to me like) is `node-gyp`. Here are the urls for the relevant repos. You might want to check out some of the README files except where it's a wiki page???. I am _not_ saying this list is either minimal or exhaustive. Not by _any_ means:

[https://github.com/felixrieseberg/windows-build-tools](url)
[https://github.com/xtermjs/xterm.js](url)
[https://github.com/xtermjs/xterm.js/wiki/Contributing](url)
[https://github.com/nodejs/node-gyp](url)
[https://github.com/Microsoft/nodejs-guidelines/](url)
[https://github.com/Microsoft/node-pty/](url)
[https://github.com/nodejs/node-gyp](url)

I didn't know of xterm.js. It's awesome. What struck my interest especially is that it's written in typescript.

Please be gentle with me. I am genuinely doing my best. Thank you all for all your patience so far and in the future.

I'm signing off now for at least 12 hours
Mike

**Comment by mikereape** [comment]
Hi sorry I missed one:

[https://github.com/microsoft/node-pty](url)

Not so sure about this one. I suspect you already cover it???

**Comment by wsw0108** [comment]
only https://github.com/felixrieseberg/windows-build-tools is needed to install app “terminal“(specially for it's deps node-pty).

https://github.com/Microsoft/node-pty/#windows

**Comment by mikereape** [comment]
Yes I knew that. I like my terminal tho as an old school old timer myself :) I did try to install the build dependencies but the install failed. I'll try again when I get time. Maybe about 12 hours from now. Thx for the confirmation. Best, Mike

**Comment by mikereape** [comment]
Hi

First, my apologies to all for being AWOL for 15 days! I have been hard at work trying to resolve my issues. I believe I have found a way forward which will at least serve my purposes though maybe not the goals of the project more generally. Before I post again however I'd like to ask what I hope is a simple question. Above, it was suggested that I try to run the four tasks in the app directory and three succeeded while the terminal one did not.  It looked to me after many attempts that the result was not something that made it possible to use those three apps because of the way the "build process" works. Is that roughly correct? Given the logs I posted previously, is there something relatively straightforward I could do to let it run to completion without the terminal build step? This would allow me to try to progress on my own and hopefully contribute back. A significant amount of effort my end will be needed even just to document what I believe is the way forward even if I don't try to make it work. To summarise, all I need are some clear answers (the simpler the better!) to the questions above.

Thank you all for all your efforts so far revolving around this issue, for sticking with me and for not closing the issue. I feel very encouraged.

I look forward to a reply
Best regards, Mike

**Comment by MatthewZMD** [comment]
Hi, sorry for the late reply, been very busy lately.

> Above, it was suggested that I try to run the four tasks in the app directory and three succeeded while the terminal one did not.

If three succeeded, then you *should* be able to use these three except the EAF Terminal application. Have you tried it?

> It looked to me after many attempts that the result was not something that made it possible to use those three apps because of the way the "build process" works. Is that roughly correct?

"build process" is a very vague term to apply here, but sure, you should be able to use those three apps.


> Given the logs I posted previously, is there something relatively straightforward I could do to let it run to completion without the terminal build step?

Try the latest `install-eaf-win32.js` file to see if it works for you. There had been updates last week.

**Comment by mikereape** [comment]
Hi thanks for the reply I've been **extremely** busy myself. It'll be at least a week till I get to it. Thanks, Mike

**Comment by manateelazycat** [comment]
@mikereape Problem still? Can you help test newest version?

**Comment by laowushi** [comment]
I installed EAF on my windows 10 machine yesterday successfully.  Some tricks here:
1. You must update npm first.
2. If you have installed python2 and python3, you must remove python2 from windows Environment path.
3. You must install emacs-epc manually.

Then, run `node install-eaf-win32.js`. 

One more thing, if you want to  "EAF-open"  .org files, you should add (require 'eaf-org) to .emacs or init.el . Otherwise, it won't work.

**Comment by MatthewZMD** [comment]
Closing since there hasn't been any response for 2 months and it is tested to work on Windows.

**Comment by mikereape** [comment]
Hi Matthew

Yes I have been away. Poor health. I'm old. I have been trying out eaf cautiously and optimistically. For a couple of weeks it seemed to be working well. However, I am back to the same kind of problems I had at the beginning. Please find my most recent install log attached. 

[install error.log](https://github.com/manateelazycat/emacs-application-framework/files/6812869/install.error.log)

This is after a clean install of python 3.9.6, the VS Studio 2017 build tools and Nodejs. A clean slate. I have seen other references to this type of problem online but nothing with that specific error.

I also see that you are using node-pty along the lines of what I talked about months ago. I cannot see how eaf could ever have worked without it on Win 10 Pro from what I know about the constellation of issues surrounding winpty and conpty which I may say is considerable by now.

I am mindful of the last report of it working on Win 10 but nothing mentioned there is/was relevant to this case. I have installed and re-installed many times in the last 2-3 weeks and now all of a sudden this is happening.

I would be most grateful for any help, suggestions or pointers. I have been intensely studying eaf and how it installs/builds itself as a learning opportunity old dinosaur that I am. I however have very little time to devote to this. My health is appalling the now.

Blessings,
Mike from the auld country

**Comment by mikereape** [comment]
Hi Matthew this is very odd. Out of curiousity I tried eaf since my last comment despite the error message I got and not getting the usual "Eaf is work" or whatever it is ... and it appears to be working!??? FWIW, the last issue you closed about PyQt5 not being found I was getting along the way as well but haven't had that problem this time. It shouldn't be working. But it appears that  it is. In which case it shouldn't be generating erroneous failure messages.

Blessings, Mike
